### ISSUE TITLE
add a shortcut function for maxout?

### ISSUE THREAD

**Issue by benanne** [benanne/Lasagne#193]
Implementing maxout in Lasagne is non-obvious because it is never called by that name anywhere (see https://groups.google.com/forum/#!topic/lasagne-users/_cd-ODu6AP8 ).

Obviously we should document `FeaturePoolLayer` so people know to use this for it, but maybe we can also add a shortcut that does something like this:

```
def maxout(incoming, num_units, ds, **kwargs):
    l1a = lasagne.layers.DenseLayer(incoming, nonlinearity=None,
                                    num_units=num_units * ds, **kwargs)
    l1 = lasagne.layers.FeaturePoolLayer(l1a, ds=ds)
    return l1
```

What does everyone think?

**Comment by ebenolson** [comment]
I don't really like it.. seems like unneeded complexity especially if you're going to support convolutional layers as well. 

Why not rename or alias FeaturePoolLayer to MaxoutLayer?

**Comment by benanne** [comment]
Because you can also pool features with different pooling functions :)

If we don't have it in the library we should at least have an example in the docs (we should probably have that regardless).

**Comment by ebenolson** [comment]
yes, but the default is max.. you could just note in the docs that MaxoutLayer can be used for other pooling, instead of that FeaturePoolLayer can be used for Maxout.

**Comment by benanne** [comment]
I disagree, I think names are important and should not be unnecessarily specific.

**Comment by ebenolson** [comment]
Ok, but there is for example `Pool2DDNNLayer` and the more specific `MaxPool2DNNLayer` (which is functionally identical). 

How about adding `MaxoutLayer = FeaturePoolLayer` just to help people find it? (although I think anyone who's read the maxout paper should understand to use `FeaturePoolLayer`).

**Comment by benanne** [comment]
Fair point, I guess that would work too :)

**Comment by f0k** [comment]
I'm fine with adding macro functions like these. It should probably be called `dense_maxout()`, though, to highlight that it's creating a dense layer of maxout units. We could have similar macro functions for convolutional layers then.

It would also be nice to have a more generic `maxout()` helper function, but that would probably look very ugly then. Something like:
```python
def maxout(layer_class, ds, axis=None, **kwargs):
    if "num_units" in kwargs:
        kwargs["num_units"] *= ds
    elif "num_filters" in kwargs:
        kwargs["num_filters"] *= ds
    else:
        raise ValueError("Unsupported layer class %r." % (layer_class.__name__))
    layer = layer_class(**kwargs)
    if hasattr(layer, 'nonlinearity'):
        layer.nonlinearity = lasagne.nonlinearities.identity
    if axis is None:
        axis = 0 if isinstance(layer, c01bsomething) else 1
    return FeaturePoolLayer(layer, ds, axis)
```
You would need to call this like `maxout(DenseLayer, 2, num_units=50)` then, but it would avoid creating a bunch of similarly-looking macro functions for different classes.

**Comment by f0k** [comment]
Comment overlap. Aliasing `MaxPoolLayer` to `FeaturePoolLayer` and documenting its use for maxout in the docstring would be equally fine, I guess, and much simpler.

**Comment by etienne87** [comment]
Hello! 
thank you for bringing lasagne & this feature.
i have just read the paper and maxout sounds like a very nice idea. There is just one thing that i'm not entirely sure... 
- when not to use it? in example above you use DenseLayer, does it mean it's a bad idea after a convLayer?
- compared to a layer using simple reLu, do you need to multiply the number of units * ds to keep a network with equivalent capacity?

**Comment by benanne** [comment]
* you can also use maxout for convolutional layers.
* yes and no - each individual maxout unit will be more powerful, so you might not need to multiply the number of linear units by ds (maybe a smaller factor). It depends on the problem.

**Comment by etienne87** [comment]
Thanks a lot for your answer!  
1. Is it planned to have n-dim pooling (so that maxout would be more an application than a real function)? 
2. Can p-norm with 3dim kernel & stride >1 achieve same result ? (http://arxiv.org/pdf/1412.6806v3.pdf)

**Comment by benanne** [comment]
1. not sure what you mean with the phrase "more an application than a real function". Also n-dim pooling is not really unambiguously defined. There is a MaxPool2DLayer currently, but I guess there is no reason why you couldn't have a MaxPoolLayer that just adapts to the dimensionality of the input. Of course for maxout you want to pool over the channel dimension, not the topological dimensions of the individual examples (i.e. spatial, temporal, ...).

2. When linking arxiv articles, linking the abstract page is much more convenient than linking to the PDF directly. Just a pet peeve of mine :)

The p-norm pooling described in that paper is spatial, afaik, not across channels, so it is not related to maxout. p-norm pooling is a generalization of max-pooling: for p=1 you get mean pooling, for p towards infinity you get max pooling.

**Comment by etienne87** [comment]
hello! once again thanx for your answer!

1. what i tried to say was that maxout would be a special use-case of 3D pooling with kernel size **(ds,1,1)**, but could take another value **(ds,3,3)** , thus achieving maxout & spatial pooling at the same time. 

2. Thanks for the advise! i will watch out now :-p ! You are right, p-norm mimics max or min (if P << 0). Let's pretend that you would put a 3D convolution somewhere in your network, instead of 2d, on a blob data of (channels, height, width). it means you would have less parameters to optimize? (since you do not take all channels at once at each coordinate (x,y) ), but still output many features, and, with p-norm...sort of mimic both maxout and 2d max pooling ?

**Comment by benanne** [comment]
1. That would work, yes. Although if we implement a 3D pooling layer it will probably expect a 5D input tensor, so that would be of no use to pool across the channel dimension. You'll need a custom layer implementation for this.

2. It doesn't really make sense to convolve along the channel dimension though, it has no topological structure.

**Comment by etienne87** [comment]
1. ok! is automatic reshape a bad thing? like reshape in (n,cn,h,w) -> (n,1,cn,h,w) / maxpool3d / reshape back 
2. fair enough i didn't think of it that way....even if correctly reordered?
http://cs.nyu.edu/~yann/research/sparse/group-psd-filters.png

**Comment by martinxsliu** [comment]
I don't think this approach of using the FeaturePoolLayer *exactly* implements the Maxout layer as described in the original paper by Goodfellow et al? The paper calls for the use of a separate bias parameter for each feature map input while FeaturePoolLayer just uses the one bias in the incoming dense layer.

**Comment by benanne** [comment]
I'm not sure I understand what you mean with that last sentence. If you use `DenseLayer` + `FeaturePoolLayer`, each linear 'component' feature of each maxout feature has its own independent bias parameter. That's how it's supposed to be, right?

**Comment by martinxsliu** [comment]
My understanding according to the paper is that the Maxout layer with k 'components' is like having k `DenseLayer`s that take in the same input and having same number of units, with the output being the maximum value of each respective unit across the k layers. Therefore there are k biases associated with each pseudo-layer.

If we use a single `DenseLayer` with k times as many units and then aggregate using the `FeaturePoolLayer`, we will only have the 1 bias unit.

Does this make sense or is there a gap in my understanding somewhere?

**Comment by benanne** [comment]
The latter, I'm afraid :) A dense layer always has a separate bias parameter for each unit. I think you may be confused by the idea of adding a 'bias unit' as input to a layer that is always equal to +1, so that the biases can be incorporated into the weight matrix of the layer. But it's important to differentiate between units and parameters in that case (and just in general, really).

**Comment by martinxsliu** [comment]
Oh you are right. I was getting the two muddled up. Thanks for clearing that up :)
